File: lee.cpp
#include <iostream>
#include <random>
#include <string>
#include <queue>
#include <algorithm>
#include <limits>

using namespace std;

// struct for node
struct node {
	int row;
	int col;
	int dist;
	int type;
};

// random number generator
mt19937 rng;

// size of graph (SIZE x SIZE matrix)
const int SIZE = 40;
// uniform random distribution for generating start and finish coords
uniform_int_distribution<mt19937::result_type> runif_SIZE(0, SIZE - 1);
// graph
node M[SIZE][SIZE];
// pointer to start and finish nodes
node *start;
node *finish;
// queue of nodes to expand
queue<node*> q;

// randomly populate the graph with start, finish and obstacles
// (may or may not be solvable)
void gen_matrix() {
	
	// arrays for start and finish coordinates
	int start_coords[2];
	int finish_coords[2];
	
	// randomly generate start coordinates
	for (int i = 0; i < 2; i++) {
		start_coords[i] = runif_SIZE(rng);
	}
	// generate finish coordinates and ensure they are not the same
	do {
		for (int i = 0; i < 2; i++) {
			finish_coords[i] = runif_SIZE(rng);
		}
	} while (start_coords[0] == finish_coords[0] && start_coords[1] == finish_coords[1]);
	
	// populate graph
	for (int i = 0; i < SIZE; i++) {
		for (int j = 0; j < SIZE; j++) {
			M[i][j] = { i, j, 0, 0 };
			// radomly make node an obstacle
			if (runif_SIZE(rng) < 4) {
				M[i][j].type = 3; // obstacle
			}
			// set node as start if coordinates match
			if (i == start_coords[0] && j == start_coords[1]) {
				M[i][j].type = 2; // start
				start = &M[i][j];
			}
			// set node as finish if coordinates match
			if (i == finish_coords[0] && j == finish_coords[1]) {
				M[i][j].type = 1; // finish
				finish = &M[i][j];
			}
		}
	}
}

// print graph to console
void print_matrix() {
	
	// loop through row indices
	for (int i = 0; i < SIZE; i++) {
		// loop through column indices
		for (int j = 0; j < SIZE; j++) {
			// element to print
			string element = "";
			// type of node
			int type = M[i][j].type;
			// if normal node
			if (type == 0) {
				int dist = M[i][j].dist;
				if (dist >= 0 && dist <= 9) {
					element += " ";
				}
				// element to print is dist (distance from start)
				element += to_string(dist);
			}
			// print F for finish node
			else if (type == 1) {
				element += " F";
			}
			// print S for start node
			else if (type == 2) {
				element += " S";
			}
			// print # for obstacles
			else if (type == 3) {
				element += " #";
			}
			// print . for nodes on path
			else if (type == 4) {
				element += " .";
			}
			cout << element << "  ";
		}
		cout << "\n" << endl;
	}
	cout << endl;
}

// returns vector of (pointers to) all non-obstacle nodes adjacent to given
vector<node*> get_adj_nodes(const node * const n) {

	vector<node*> adj_nodes = {};
	
	// row and column indices of current node
	int row = n->row;
	int col = n->col;
	// row index - 1 (one row north)
	int north = n->row - 1;
	// row index + 1 (one row south)
	int south = n->row + 1;
	// column index - 1 (one column west)
	int west = n->col - 1;
	// column index + 1 (one column east)
	int east = n->col + 1;
	
	// for NSWE, add (pointer to) adjacent node to vector if
	// it is valid (within bounds of graph) and is either 
	// a standard node or the finish node
	if (north >= 0 && (M[north][col].type <=2)) {
		adj_nodes.push_back(&M[north][col]);
	}
	if (south <= SIZE - 1 && (M[south][col].type <= 2)) {
		adj_nodes.push_back(&M[south][col]);
	}
	if (west >= 0 && (M[row][west].type <= 2)) {
		adj_nodes.push_back(&M[row][west]);
	}
	if (east <= SIZE - 1 && (M[row][east].type <= 2)) {
		adj_nodes.push_back(&M[row][east]);
	}
	
	// return vector of (pointers to) adjacent nodes
	return adj_nodes;
}

int iterations = 0;

void find_path() {
	
	q.push(start);
	bool done = false;
	while (!done) {
		iterations++;
		node *next_node = q.front();
		q.pop();
		for (node *n : get_adj_nodes(next_node)) {
			if (n->type == 1) {
				done = true;
			}
			else if (n->type <= 1 && n->dist == 0) {
				n->dist = max(next_node->dist + 1, 1);
				q.push(n);
			}
		}
	}
}

void backtrace() {
	
	int min_dist = numeric_limits<int>::max();
	node *pos = finish;
	bool done = false;
	while(!done) {
		for (node *n : get_adj_nodes(pos)) {
			if(n->type == 2) {
				done = true;
			}
			else if (n->dist < min_dist && n->dist > 0) {
				min_dist = n->dist;
				pos = n;
			}
		}
		if(!done) {
			pos->type = 4;
		}
	}	
}

int main() {
	
	rng.seed(random_device()());
	
	gen_matrix();
	print_matrix();
	find_path();
	print_matrix();
	backtrace();
	print_matrix();
	cout << "Iterations: " << iterations << endl;
	
	return 0;
}

File: lee.py

class queue:
    
    def __init__(self):
        self.q = []
    
    def enqueue(self, x):
        self.q.append(x)
        
    def pop(self):
        i = self.q[0]
        del(self.q[0])
        return i
    
    def get_q(self):
        return self.q

A = []

for i in range(20):
    A.append([])

for i in range(20):
    for j in range(20):
        A[i].append([i, j, 0])
        
A[12][2][2] = -1
A[3][4][2] = -2
        
def get_adj_coords(x, y):
    
    l = []
    
    if y+1 in range(20):
        l.append([x, y+1])
    if y-1 in range(20):
        l.append([x, y-1])
    if x+1 in range(20):
        l.append([x+1, y])
    if x-1 in range(20):
        l.append([x-1, y])
    
    return l

def print_matrix():
    for i in range(len(A)):
        for j in range(len(A[0])):
            el = A[i][j][2]
            if el in range(10):
                el = " " + str(el)
            else:
                el = str(el)
            print(el + " ", end='')
        print('\n')

start = [12, 2]
end = [3, 4]

q = queue()

for i in get_adj_coords(*start):
    x, y = i
    A[x][y][2] = 1
    q.enqueue(A[x][y])
    
done = False
while not done:
    next_cell = q.pop()
    for i in get_adj_coords(next_cell[0], next_cell[1]):
        x, y = i
        if A[x][y][2] == -2:
            done = True
        elif A[x][y][2] == 0:
            A[x][y][2] = next_cell[2] + 1
            q.enqueue(A[x][y])
            print(A[x][y])
            
next_coords = end
done = False
while not done:
    min_dist = 1000
    coords_min = []
    for i in get_adj_coords(next_coords[0], next_coords[1]):
        x, y = i
        if A[x][y][2] == -1:
            print("Done")
            done = True
        elif A[x][y][2] < min_dist and A[x][y][2] > 0:
            min_dist = A[x][y][2]
            coords_min = [x, y]
    if not done:
        next_coords = coords_min
        print(next_coords)
        A[next_coords[0]][next_coords[1]][2] = -3
    
        
    
        


